### ISSUE TITLE
Child templates and helper data

### ISSUE THREAD

**Issue by zheffie** [jsreport/jsreport#52]
I'm using helpers to request data to be rendered through an ajax call. 
However it seems that this data is not passed on to the child templates.

Is this expected/desired behaviour?

**Comment by pofider** [comment]
With child templates you mean [child templates extension](http://jsreport.net/learn/child-templates)?
An with helpers you mean the functions you specify in the helpers tab resp. template.helpers property?

**Comment by zheffie** [comment]
Yes to both questions :)

**Comment by pofider** [comment]
The helper functions are synchronous, so I don't understand how you can use it for AJAX calls.

**Comment by zheffie** [comment]
Sorry,  my mistake.  The Ajax call is in a script

**Comment by pofider** [comment]
Ok, this is quite likely not desired behavior.

The problem is that rendering pipeline processes child templates before the script execution. This sounds to me right now like a bug and I will keep this open until I make my mind about it.

However you can now trick it with the simple workaround. The child template extraction is actually processed twice in the pipeline. At the beginning of the request and then again after the templating engines are rendered. This means you can construct the child template name by simple helper function and postpone child template extraction after the script loads data.

In the template you put something like
```html
{#child {{>~getName()}}}
```

and add the helper
```js
function getName() {
  return "myChildName";
}
```

It would be probably the best if the child templates would be always extracted just after the templating engines are executed and that is it. It would keep possible to assembly child template name dynamically as well as use scripts to fill the data. 

If anyone has an opinion on this, please share it with me.

**Comment by zheffie** [comment]
Thanks for the input. I tried the proposed mechanism, but doesn't seem to work
I added the helper (returns the template name), and changed the template but I get the following error:

Error occured - Error during rendering report: Parse error on line 72:
.../table>
{#child {{>~getName()}}%}%

<
----------------------^
Expecting 'OPEN_SEXPR', 'ID', 'STRING', 'NUMBER', 'BOOLEAN', 'DATA', got 'INVALID'
Stack - Error: Parse error on line 72:
.../table>
{#child {{>~getName()}}%}%

**Comment by pofider** [comment]
My proposal works for jsrender templating engine. For handlebars it should be:
```html
{#child {{childName}}}
```

**Comment by zheffie** [comment]
Using "{#child {{getName}}}", I get "[object Object]" rendered in the pdf.

**Comment by pofider** [comment]
Is the child having the html recipe set?

**Comment by zheffie** [comment]
yes
the child template:
![image](https://cloud.githubusercontent.com/assets/4513400/6847475/585ded12-d3c7-11e4-8b25-6fec8053f6a0.png)

the main template:
![image](https://cloud.githubusercontent.com/assets/4513400/6847490/74c5c4b6-d3c7-11e4-932e-ae1f22760813.png)

**Comment by pofider** [comment]
Ups. Sorry.

Please wait until the v0.3 which is already in phase of merging testing now. There it will work.

**Comment by zheffie** [comment]
Np. Thanks

**Comment by zheffie** [comment]
Any estimate on the 0.3 release date?

Thanks in advance

**Comment by pofider** [comment]
We are almost there... I hope in the next hours it's released.

**Comment by pofider** [comment]
Now it is released!

**Comment by zheffie** [comment]
Is the "request.data" object still available in the scripts? I'm getting errors after upgrading to 0.3

**Comment by pofider** [comment]
Yes, and I tested your scenario quite precisely so I am wondering what could went wrong.

I have a script like this and it is able to push data into the child templates.
```js
function beforeRender(done) {
    setTimeout(function() {
        request.data = {
            "foo": "xxxxx"
        }
        done();    
    }, 5000)
}
```

Maybe you can try to remove the jsreport folder and install it again if it helps...

**Comment by RyanThomas** [comment]
How do I unsubscribe?

On Mon, Mar 30, 2015 at 2:43 PM, Jan Blaha <notifications@github.com> wrote:

> Yes, and I tested your scenario quite precisely so I am wondering what
> could went wrong.
>
> I have a script like this and it is able to push data into the child
> templates.
>
> function beforeRender(done) {
>     setTimeout(function() {
>         request.data = {
>             "foo": "xxxxx"
>         }
>         done();
>     }, 5000)
> }
>
> Maybe you can try to remove the jsreport folder and install it again if it
> helps...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsreport/jsreport/issues/52#issuecomment-87786935>.
>



-- 
Ryan Thomas, MCP
University of Cincinnati | DAAP

Please consider the environment before printing this email.

**Comment by zheffie** [comment]
Still confused.

Problem 1:

In 0.2.10, I was able to this

script:
var req=require('request');
function beforeRender(done) {
  req({url:'http://www.url.com?id='+request.data.id,json:true},function(err,response,body){
    request.data.items=response.body;
    done();
  });
}

data:
data = {id:1234}

This no longer seems to work.

problem 2:
Following your example: whatever ends up in request.data does not propagate to the child template.

**Comment by zheffie** [comment]
Problem 2 is solved:

```
{#child mySnippet}
```

mySnippet is a template with `html` recipe, and engine set to `none`

**Comment by zheffie** [comment]
Problem 1 is no longer a problem when calling the API directly and providing `{id:1234}` as input.
The behaviour seems inconsistent with the studio behaviour though

**Comment by pofider** [comment]
I see...

1.
The original problem was that scripts were evaluated too late in the pipeline and now the evaluation happens too early. :) 

The script extension should be evaluated just after the data are loaded using data extension, but before the childTemplates extraction happens. I need to fix and document this more precisely, but I will put it into the next maintenance release of 0.3.

Thank you for your valuable inputs.

**Comment by zheffie** [comment]
No problem. I'm trying to figure out how it works because I really see some potential in your tool. Thanks for the great work!

One more question regarding 'helpers' and 'childtemplates'. It seems that the helper functions from child templates are not recognized/parsed. I have to add the helpers functions to the parent template helpers. Is this as designed?

**Comment by pofider** [comment]
The helper functions are separated. In other words you can use parent's helpers only in the parent and child template helpers only in the particular child.

If you are struggling with something you can share don't hesitate to zip your data folder and send it to me <jan.blaha at hotmail.com> or you can also reach me on skype <pofider> for chat.

**Comment by pofider** [comment]
I guess the main problem from this issues is already solved.

Similar issue is to let child templates run scripts which will be solved in #61
